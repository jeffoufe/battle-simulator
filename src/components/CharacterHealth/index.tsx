import React from 'react';
import {
    CharacterHealthContainer,
    CharacterHealthIndicatorContainer,
    CharacterHealthIndicator,
    CharacterHealthPoints
} from './components';

interface CharacterHealthProps {
    health: number
}

const CharacterHealth = (props: CharacterHealthProps) => (
    <CharacterHealthContainer>
        <CharacterHealthIndicatorContainer>
            <CharacterHealthIndicator health={props.health} />
        </CharacterHealthIndicatorContainer> 
        <CharacterHealthPoints>{props.health}</CharacterHealthPoints>
    </CharacterHealthContainer>
)

CharacterHealth.displayName = 'CharacterHealth';
export default CharacterHealth;