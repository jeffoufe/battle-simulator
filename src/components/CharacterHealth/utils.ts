export const getHealthColor = (health: number) => {
    if (health > 70) {
        return 'green';
    }
    if (health > 30) {
        return 'orange';
    } 
    return 'red';
}