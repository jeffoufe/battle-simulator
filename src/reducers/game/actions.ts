import { 
    PLAYER_ROLL, 
    RESET_GAME,
    MONSTER_ROLL,
    END_TURN
} from './constants';

export const clickAttack = () => ({
    type: PLAYER_ROLL
})

export const resetGame = () => ({
    type: RESET_GAME
})

export const monsterRoll = (result: number) => ({
    type: MONSTER_ROLL,
    payload: { result }
})

export const endTurn = (result: number) => ({
    type: END_TURN,
    payload: { result }
})