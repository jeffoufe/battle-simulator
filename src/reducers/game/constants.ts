export const PLAYER_ROLL = 'GAME/PLAYER_ROLL';
export const MONSTER_ROLL = 'GAME/MONSTER_ROLL';
export const ALL_ROLLS_DONE = 'GAME/ALL_ROLLS_DONE';
export const RESET_TURN = 'GAME/RESET_TURN';
export const END_TURN = 'GAME/END_TURN';
export const COMPUTE_TURN = 'GAME/COMPUTE_TURN';
export const RESET_GAME = 'GAME/RESET_GAME';