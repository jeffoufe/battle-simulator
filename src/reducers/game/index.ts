import { State } from '..';
import { GameReducerAction } from './types';
import {
  PLAYER_ROLL,
  MONSTER_ROLL,
  ALL_ROLLS_DONE,
  RESET_TURN,
  COMPUTE_TURN,
  RESET_GAME
} from './constants';

export const CHARACTER_INITIAL_STATE = {
  health: 100,
  result: 0,
  isRolling: false
};

export const INITIAL_STATE = {
  isTurnActive: false,
  player: CHARACTER_INITIAL_STATE,
  monster: CHARACTER_INITIAL_STATE
}

export default (state: State = INITIAL_STATE, action: GameReducerAction) => {
  switch (action.type) {
    case PLAYER_ROLL:
      return {
        ...state,
        isTurnActive: true,
        player: {
          ...state.player,
          isRolling: true
        }
      }
    case MONSTER_ROLL:
      return {
        ...state,
        monster: {
          ...state.monster,
          isRolling: true
        },
        player: {
          ...state.player,
          isRolling: false,
          result: action.payload.result
        }
      }
    case ALL_ROLLS_DONE:
      return {
        ...state,
        monster: {
          ...state.monster,
          isRolling: false,
          result: action.payload.result
        }
      }
    case COMPUTE_TURN:
      const damage = state.player.result - state.monster.result;
      const character = damage < 0 ? 'player' : 'monster';
      return {
        ...state,
        [character]: {
          ...state[character],
          health: state[character].health - (Math.abs(damage))
        }
      }
    case RESET_TURN:
      return {
        ...state,
        isTurnActive: false,
        player: {
          ...state.player,
          result: 0
        },
        monster: {
          ...state.monster,
          result: 0
        }
      }
    case RESET_GAME:
      return INITIAL_STATE;
    default:
      return state;
  }
};