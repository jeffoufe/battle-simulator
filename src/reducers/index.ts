export { default as gameReducer } from './game';

interface Character {
    isRolling: boolean,
    health: number,
    result: number
};

export interface State {
    player: Character,
    monster: Character,
    isTurnActive: boolean
}