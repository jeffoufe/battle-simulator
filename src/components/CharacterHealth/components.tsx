import styled from 'styled-components';
import { getHealthColor } from './utils';
import { breakpoints } from '../../constants';

interface CharacterHealthIndicatorProps {
    health: number
}

export const CharacterHealthContainer = styled.div`
    display: flex;
    flex-direction: column;
`

export const CharacterHealthIndicatorContainer = styled.div`
    width: 40px;
    height: ${60}px;
    border: 1px solid black;
    position: relative;

    @media only screen and (min-width: ${breakpoints.tablet}) {
        height: 150px;
    }

    @media only screen and (min-width: ${breakpoints.laptop}) {
        height: 200px;
    }
`

export const CharacterHealthIndicator = styled.div`
    width: 40px;
    height: ${(props: CharacterHealthIndicatorProps) => props.health}%;
    background-color: ${(props: CharacterHealthIndicatorProps) => getHealthColor(props.health)};
    position: absolute;
    bottom: 0;
`

export const CharacterHealthPoints = styled.span`
    margin-top: 5px;
    margin-bottom: 0;
    text-align: center;
`