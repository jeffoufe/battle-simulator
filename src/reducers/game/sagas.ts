import { ALL_ROLLS_DONE, END_TURN, RESET_TURN, COMPUTE_TURN } from './constants';
import { takeEvery, put, select, delay } from 'redux-saga/effects';

interface EndTurnAction {
    payload: {
        result: number
    }
}

function* endTurn(action: EndTurnAction) {
    const { result } = action.payload;
    yield put({
        type: ALL_ROLLS_DONE,
        payload: { result }
    })
    const playerResult = select((state: any) => state.player.result);
    const monsterResult = select((state: any) => state.monster.result);
    if (playerResult === monsterResult) {
        yield delay(1000);
        yield put({
            type: RESET_TURN
        });
    } else {
        const order = playerResult < monsterResult;
        yield put({
            type: COMPUTE_TURN,
            payload: { order }
        })
        yield delay(1000);
        yield put({
            type: RESET_TURN
        });
    }
}   

export function* watchEndTurn() {
    // @ts-ignore
    yield takeEvery(END_TURN, endTurn);
}