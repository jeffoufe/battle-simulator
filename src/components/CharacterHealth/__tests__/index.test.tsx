import React from 'react';
import { shallow } from 'enzyme';
import CharacterHealth from '..';
import * as redux from 'react-redux';
import { INITIAL_STATE } from '../../../reducers/game';

describe('CharacterHealth', () => {
    it('No dices rolled', () => {
        const useSelectorSpy = jest.spyOn(redux, 'useSelector'); 
        useSelectorSpy.mockReturnValue(INITIAL_STATE);

        const wrapper = shallow(
            <CharacterHealth health={50} order={1} />
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('CharacterHealthPoints').at(0).text()).toEqual('');
    })

    it('Dices rolled', () => {
        const useSelectorSpy = jest.spyOn(redux, 'useSelector'); 
        useSelectorSpy.mockReturnValue({ 
            ...INITIAL_STATE,
            player: {
                ...INITIAL_STATE.player,
                result: 3
            },
            monster: {
                ...INITIAL_STATE.monster,
                result: 5
            }
        });

        const wrapper = shallow(
            <CharacterHealth health={50} order={0} />
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('CharacterHealthPoints').at(0).text()).toEqual('-2');
    })
})