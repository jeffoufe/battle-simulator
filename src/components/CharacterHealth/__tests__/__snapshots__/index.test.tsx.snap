// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterHealth Dices rolled 1`] = `
<CharacterHealthContainer>
  <CharacterHealthPoints>
    -2
  </CharacterHealthPoints>
  <CharacterHealthIndicatorContainer>
    <CharacterHealthIndicator
      health={50}
    />
  </CharacterHealthIndicatorContainer>
  <CharacterHealthPoints>
    50
  </CharacterHealthPoints>
</CharacterHealthContainer>
`;

exports[`CharacterHealth No dices rolled 1`] = `
<CharacterHealthContainer>
  <CharacterHealthPoints />
  <CharacterHealthIndicatorContainer>
    <CharacterHealthIndicator
      health={50}
    />
  </CharacterHealthIndicatorContainer>
  <CharacterHealthPoints>
    50
  </CharacterHealthPoints>
</CharacterHealthContainer>
`;
